Initial module setup

natali-zaitseva opened the pull request:
## **Purpose**
Per 
https://issues.folio.org/browse/MODKBEKBJ-14 
https://issues.folio.org/browse/MODKBEKBJ-45
https://issues.folio.org/browse/MODKBEKBJ-44
https://issues.folio.org/browse/MODKBEKBJ-16
https://issues.folio.org/browse/MODKBEKBJ-15
https://issues.folio.org/browse/MODKBEKBJ-3
we are supposed to create an initial file set for the module 
## Approach
* Generate pom and all related files using https://github.com/folio-org/mod-rmb-template 
* Create ModuleDescriptor and DeploymentDescriptor files
* Add raml samples files
* Copy raml examples from the https://github.com/folio-org/mod-kb-ebsco/tree/master/ramls/examples

dmtkachenko commented on ramls/eholdings.raml:
```diff
@@ -0,0 +1,141 @@
+#%RAML 1.0
+title: mod-kb-ebsco
```
mod-kb-ebsco-java?

natali-zaitseva replied:
you are right. Thank you, will fix it

dmtkachenko commented on ramls/eholdings.raml:
```diff
@@ -0,0 +1,141 @@
+#%RAML 1.0
+title: mod-kb-ebsco
+baseUri: https://github.com/folio-org/mod-kb-ebsco
```
mod-kb-ebsco-java?

natali-zaitseva commented on descriptors/ModuleDescriptor-template.json:
```diff
@@ -0,0 +1,124 @@
+  "name": "Module KB EBSCO Module",
+  "provides": [
+    {
+      "id": "${artifactId}",
```
Thank you Kateryna

sduvvuri-ebsco commented on .gitignore:
```diff
@@ -21,3 +21,12 @@
+/src/main/java/org/folio/rest/client/*
+/src/main/java/org/folio/rest/jaxrs/*
+/src/main/java/storage/model/*
+/src/main/resources/postgres-conf.json
```
Sorry - unsure what .idea and *.iml are but since this is a backend module - we probably do not need node_modules. We also might not need storage/model/* since this module does not talk to DB, same with /resources/postgres-conf.json - what do you think?

We might want to add

natali-zaitseva replied:
Hello Sobha,
.idea and .iml files are specific files for IDE that we use
and you are right we can get rid of node_modules, storage/model/ and resources/postgres-conf.json. 
Thank you for the review

sduvvuri-ebsco replied:
👍

sduvvuri-ebsco commented on Jenkinsfile:
```diff
@@ -0,0 +1,14 @@
+    buildJavaDocker {
+      publishMaster = 'yes'
+      healthChk = 'yes'
+      healthChkCmd = 'curl -sS --fail -o /dev/null  http://localhost:8081/admin/health || exit 1'
```
If we want to have this as healthChkCmd, we might want to add a user story to implement `/admin/health` endpoint, otherwise this health check will always fail

sduvvuri-ebsco commented on NEWS.md:
```diff
@@ -0,0 +1,10 @@
+ * Updated README file
+ * Added boilterplate files
+## 2018-10-04 v1.0.0
+ * Repository creation
```
I am not sure if there is a specific pattern for publishing news - but when I check mod-notes as an example, they have version first and then date - not sure if this matters though - also observed that versions start off with 0.1.0 and go all the way up based on minor and major updates

carolegodfrey replied:
Was also wondering if version should be a 0.1.0 initially -- not sure of best practice for this

dmtkachenko replied:
format changed. version set to 0.1.0

sduvvuri-ebsco commented on README.md:
```diff
@@ -1,2 +1,125 @@
+ | PUT      | /eholdings/root-proxy                        | Update the current root proxy.                                                        |
+ | GET      | /eholdings/configuration                     | Get KB configuration currently being used.                                            |
+ | PUT      | /eholdings/configuration                     | Update the currently set KB configuration.                                            |
+ | GET      | /eholdings/status                            | Retrives status of current KB configuration.                                          |
```
We probably do not want to declare the API in README. We'll have it in module descriptor as well as it will be published https://dev.folio.org/reference/api/ after we have RAML in place

natali-zaitseva replied:
@sduvvuri-ebsco We added API to README file to have more visibility what requests can be handled by the module.
If this violates the privacy policy, please let me know,  and we will remove it.

natali-zaitseva replied:
ok, will remove it.

sduvvuri-ebsco commented on README.md:
````diff
@@ -1,2 +1,125 @@
+    -H "Content-type: application/json"   \
+    -d @target/TenantModuleDescriptor.json \
+    http://localhost:9130/_/proxy/tenants/<tenant_name>/modules
+```
````
Love this description - I think it will be very helpful for newbies to the project 👍

carolegodfrey commented on README.md:
````diff
@@ -1,2 +1,125 @@
+    http://localhost:9130/_/proxy/tenants/<tenant_name>/modules
+```
+
+
````
Nice details on the readme -- very helpful reference for developers!

sduvvuri-ebsco commented on descriptors/ModuleDescriptor-template.json:
```diff
@@ -0,0 +1,124 @@
+{
+  "id": "${artifactId}-${version}",
+  "name": "Module KB EBSCO Module",
```
Can we rename this to "EBSCO KB Broker" or probably an even better name? Thanks!

natali-zaitseva replied:
yes, sure. will rename it

natali-zaitseva commented on descriptors/ModuleDescriptor-template.json:
```diff
@@ -0,0 +1,124 @@
+          "methods": ["GET"],
+          "pathPattern": "/eholdings/status"
+        }
+      ]
```
@carolegodfrey may you point me to this endpoint, please? I have used this raml file https://github.com/folio-org/mod-kb-ebsco/blob/master/ramls/eholdings.raml for ModuleDescriptor creation and maybe missed something.

carolegodfrey replied:
Sure -- I was comparing to the module descriptor for mod-kb-ebsco at https://github.com/folio-org/mod-kb-ebsco/blob/master/ModuleDescriptor.json

natali-zaitseva replied:
ok, I see. I was looking for already used endpoints in the raml.  If you do not mind we will add an endpoint for custom-labels when it is in use.

sduvvuri-ebsco commented on descriptors/ModuleDescriptor-template.json:
```diff
@@ -0,0 +1,124 @@
+        },
+        {
+          "methods": ["GET"],
+          "pathPattern": "/eholdings/providers/{provider_id}"
```
Can we please maintain consistency here: Few endpoints have camelcase like "packageId" where as others have snake case like "provider_id"? I know, its our bad when we put RAML together for mod-kb-ebsco since multiple developers worked on putting together that RAML file. Thanks!

natali-zaitseva replied:
@sduvvuri-ebsco will rename it. Thanks

sduvvuri-ebsco commented on descriptors/ModuleDescriptor-template.json:
```diff
@@ -0,0 +1,124 @@
+        },
+        {
+          "methods": ["GET"],
+          "pathPattern": "/eholdings/titles"
```
we are missing POST method of /eholdings/titles

sduvvuri-ebsco commented on descriptors/ModuleDescriptor-template.json:
```diff
@@ -0,0 +1,124 @@
+        },
+        {
+          "methods": ["POST"],
+          "pathPattern": "/eholdings/titles/{title_id}"
```
This endpoint does not exist , its POST to collection `/eholdings/titles` instead of this one

natali-zaitseva replied:
thanks, will change this

sduvvuri-ebsco commented on descriptors/ModuleDescriptor-template.json:
```diff
@@ -0,0 +1,124 @@
+        }
+      ]
+    }
+  ],
```
Do we need this `_tenant` part?

natali-zaitseva replied:
yes, it is a system interface. Its purpose is to do whatever initialization the module needs, for example, to create database tables.

sduvvuri-ebsco replied:
Mmm..since this module does not need any initialization, do we still need it?

natali-zaitseva replied:
agree, will delete it.

sduvvuri-ebsco commented on descriptors/ModuleDescriptor-template.json:
```diff
@@ -0,0 +1,124 @@
+      ]
+    }
+  ],
+  "permissionSets" : [],
```
We have been pretty relaxed with `permissionSets` and `requires` in mod-kb-ebsco but I think we should put them together here. Good example that we can refer to : https://github.com/folio-org/mod-codex-ekb/blob/master/descriptors/ModuleDescriptor-template.json
Its ok if we do not want to address it in this PR - but if thats the case, can we please create a user story in the backlog so it does not slip through cracks? Thanks!

natali-zaitseva replied:
sure, @sduvvuri-ebsco. I will create a user story for adding permissions tomorrow.

sduvvuri-ebsco replied:
Thanks @natali-zaitseva!

dmtkachenko replied:
there is a MODKBEBSCO-14 for permissions. I'll discuss with Khalilah if it's the right one

sduvvuri-ebsco commented on docker/docker-entrypoint.sh:
```diff
@@ -0,0 +1,20 @@
+else
+   exec java -jar ${VERTICLE_HOME}/module.jar "$@"
+fi
+
```
We do not need this docker/docker-entrypoint.sh anymore. Added it in comments section to https://issues.folio.org/browse/MODKBEKBJ-44

sduvvuri-ebsco commented on pom.xml:
```diff
@@ -3,7 +3,7 @@
+  <version>0.1.0-SNAPSHOT</version>
   <packaging>jar</packaging>
 
   <licenses>
```
While this pom.xml is pretty concrete, there are few interesting tags we can grab from https://github.com/folio-org/mod-codex-ekb/blob/master/pom.xml like name, url, description, organization, issueManagement etc. Also, noticed some differences in <scm> with extensions, <licenses> and so on...Please compare and use your judgement to adjust this accordingly. Thanks!

natali-zaitseva replied:
we have used https://github.com/folio-org/mod-rmb-template which creates pom file as well as other boilerplate code. I have added the following tags like ciManagement, issueManagement but the other dependencies seem module-specific and we can add them when it is needed.

sduvvuri-ebsco replied:
Sounds good!

sduvvuri-ebsco commented on scripts/lint-raml-cop.sh:
```diff
@@ -0,0 +1,61 @@
+  echo "${help_msg}"
+fi
+
+exit ${result}
```
We do not need scripts/lint-raml-cop.sh either..added it to comments of https://issues.folio.org/browse/MODKBEKBJ-44. As mentioned in https://dev.folio.org/guides/raml-cop/, we should be able to run raml-cop locally without this script in place.

natali-zaitseva replied:
removed

sduvvuri-ebsco commented on ramls/providers.raml:
```diff
@@ -0,0 +1,160 @@
+          400:
+            description: Bad Request
+            body:
+              application/vnd.api+json:
```
On all the ramls/*.raml -> I am not sure if this actually builds using RMB -> did you give it a shot? When we put this RAML together using 1.0, since it is a Ruby module, it does not go through RMB and this RAML's mere purpose was to generate API documentation as we have it here: https://s3.amazonaws.com/foliodocs/api/mod-kb-ebsco/eholdings.html
If this is building through RMB and generating POJO's as expected, then well and good, otherwise, we'll have to revisit this.

sduvvuri-ebsco commented:
@dcrossleyau : Thanks for your feedback! I think Natalia removed the healthChkCmd addressing one of my comments! We'll add it back differently than earlier! Thanks once again

sduvvuri-ebsco commented on Jenkinsfile:
```diff
@@ -0,0 +1,13 @@
+      healthChk = 'yes'
+    }
+  }
+}
```
@natali-zaitseva : Lets add the healthChkCmd back in here except that we'll use a different endpoint like the below:
`healthChkCmd = 'curl -sS --fail -o /dev/null http://localhost:8081/apidocs/ || exit 1'`. Per https://github.com/folio-org/raml-module-builder, "The maven-resources-plugin in pom.xml will copy the RAML files into a directory under /apidocs so that the runtime framework can pick it up and display html documentation based on the RAML files." So, this endpoint can work as our healthChkCmd. What do you think?

sduvvuri-ebsco commented:
@natali-zaitseva : Looks pretty good for the most part. Few comments:
1. We still have to consolidate various ramls/*.raml into ramls/eholdings.raml per our discussion this morning.
2. Assuming that the work with the existing raml we have here is incomplete and that will be tweaked as we work on each endpoint to suit what RMB needs to generate interfaces and POJOs.
3. Could you please add .editorConfig and PULL_REQUEST_TEMPLATE.md from this repo to this PR? https://github.com/folio-org/mod-codex-ekb
4. Could you please add all ticket numbers that go into this PR in headline/title and fill in details in the pull request as you see fit? Similar to this PR for example: https://github.com/folio-org/mod-kb-ebsco/pull/211 - This PR might not have screenshots but certainly Purpose, Approach, Learning and TODOs and Open Questions if any can be noted on the PR notes.

Thanks so much in advance!

natali-zaitseva commented:
@sduvvuri-ebsco 
I have updated the PR and added PULL_REQUEST_TEMPLATE.md  and we already have .editorConfig file which can be found in the root folder. Regarding the ralm files - I have double checked it and we are good on that, this means we are not violating the defined contract for eholding module. As you can see from the following screenshots we are handling the endpoints with respect.
![image](https://user-images.githubusercontent.com/37537790/46796106-79d8d300-cd54-11e8-9c37-99aef4cd05b2.png)
![image](https://user-images.githubusercontent.com/37537790/46796113-7fceb400-cd54-11e8-83eb-04bdcaec0f93.png)

sduvvuri-ebsco commented:
Thanks Natalia! Looks pretty good! 👍

julianladisch commented on ramls/titles.raml:
```diff
@@ -0,0 +1,181 @@
+          body:
+            application/vnd.api+json:
+              example: !include examples/titles/titles_titleId_get_404_response.json
+    post:
```
This `post` section should be moved from `/eholdings/titles/{title_id}` to `/eholdings/titles` as correctly declared in `descriptors/ModuleDescriptor-template.json`.
